components/prism-odin.js

```diff
@@ -0,0 +1,90 @@
+(function (Prism) {
+	var escapes = /\\(?:["'\\abefnrtv]|0[0-7]{2}|U[\dA-Fa-f]{6}|u[\dA-Fa-f]{4}|x[\dA-Fa-f]{2})/.source;
+
+	function withEscapes(pattern, flags) {
+		return RegExp(pattern.replace(/<escapes>/g, escapes), flags);
+	}
+
+	Prism.languages.odin = {
+		/**
+		 * The current implementation supports only 1 level of nesting.
+		 *
+		 * @author Michael Schmidt
+		 * @author edukisto
+		 */
+		'comment': [
+			{
+				pattern: /\/\*(?:[^/*]|\/(?!\*)|\*(?!\/)|\/\*(?:\*(?!\/)|[^*])*(?:\*\/|$))*(?:\*\/|$)/,
+				greedy: true
+			},
+			{
+				pattern: /#![^\n\r]*/,
+				greedy: true
+			},
+			{
+				pattern: /\/\/[^\n\r]*/,
+				greedy: true
+			}
+		],
+
+		/**
+		 * Should be found before strings because of '"'"- and '`'`-like sequences.
+		 */
+		'char': {
+			pattern: withEscapes(/'(?:<escapes>|[^\n\r'\\])'/.source),
+			greedy: true,
+			inside: {
+				'symbol': withEscapes(/<escapes>/.source)
+			}
+		},
+
+		'string': [
+			{
+				pattern: /`[^`]*`/,
+				greedy: true
+			},
+			{
+				pattern: withEscapes(/"(?:<escapes>|[^\n\r"\\])*"/.source),
+				greedy: true,
+				inside: {
+					'symbol': withEscapes(/<escapes>/.source)
+				}
+			}
+		],
+
+		'directive': {
+			pattern: /#\w+/,
+			alias: 'property'
+		},
+
+		'number': /(?:\b0(?:b[01_]+|d[\d_]+|h_*(?:(?:(?:[\dA-Fa-f]_*){8}){1,2}|(?:[\dA-Fa-f]_*){4})|o[0-7_]+|x[\dA-F_a-f]+|z[\dAB_ab]+)\b|(?:\b\d+(?:\.\d*)?|\B\.\d+)(?:[Ee][+-]?\d*)?[ijk]?)/,
+
+		'boolean': /\b(?:_|false|nil|true)\b/,
+
+		'keyword': /\b(?:asm|auto_cast|bit_set|break|case|cast|context|continue|defer|distinct|do|dynamic|else|enum|fallthrough|for|foreign|if|import|in|map|matrix|not_in|or_else|or_return|package|proc|return|struct|switch|transmute|typeid|union|using|when|where)\b/,
+
+		'procedure': {
```

Comment:
It would be nice if we could also find procedure definitions.

![image](https://user-images.githubusercontent.com/20878432/146642011-c11cdea2-d45a-4a48-8752-594bf95bfccd.png)

Could please add the following token before `keyword`?

```js
'procedure-definition': {
  pattern: /\b\w+(?=\s*::\s*proc\b)/,
  alias: ['procedure', 'function']
}
```

Also, is `\s` ok or does it have to be `[ \t]` here as well?

Reply:
Yes, sure! I should have done this already.

`\s` is too universal everywhere in Odin. The specification [defines](https://odin-lang.org/docs/spec/#tokens) only four whitespace characters. Moreover, the current implementation forbids newlines before the first colon for some reason.

So the most consistent implementation is

```odin
\b\w+(?=[ \t]*:[ \n\r\t]*:[ \n\r\t]*proc\b)
```

or

```odin
\b\w+(?=[ \t]*(?::[ \n\r\t]*){2}proc\b)
```

I’m not sure if this is intentional and not a bug. Maybe we should ask @gingerBill for advice.

Reply:
Wow. There can [spaces between colons](https://github.com/odin-lang/Odin/blob/0548db423067bce16d45af651819bf56feb5d411/src/tokenizer.cpp#L1261). I did not expect that. I take back my comment about `::` having to be one token.

The second pattern you suggested looks good :)

Reply:
Is there a need to distinguish between declarations and calls? See cd99873, please.

And I think it is better to find procedure names **after** searching for keywords, but before Booleans (including `nil`) because the latter ones can be function names. `_` can’t. Wow.

Unfortunately, `proc` had to be moved to a separate block.

Reply:
> Is there a need to distinguish between declarations and calls?

Yes, there is. As you saw yourself, the procedure definition pattern needs to be able to see the `proc` keywords, so it has to be matched before the `proc` keyword is. However, the call syntax overlaps with some keyword, so the call pattern has to be after **all** keywords.

Right now, `cross :: proc(a, b: Vector3) -> Vector3` will not be highlighted correctly. (Please add this snippet to the procedure test.)

If the procedure definition pattern is matched before all keywords and the procedure call is matched after all keywords, then everything will be highlighted correctly.

I know that it's a little weird to split the procedure patterns, but it's necessary from a technical standpoint. We also have to do this in a few other languages, so it's not like Odin will be a first in that regard :)